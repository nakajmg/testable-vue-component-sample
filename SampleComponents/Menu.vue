<template>
  <nav>
    <div v-show="menuOpened">
      <router-link to="/">Root</router-link>
      <router-link to="/foo">Foo</router-link>
      <router-link to="/bar">Bar</router-link>
    </div>
    <button @click="toggleMenu">
      Open Menu
    </button>
  </nav>
</template>

<script>
export default {
  name: "Menu",
  data() {
    return {
      menuOpened: false,
    }
  },
  methods: {
    toggleMenu() {
      this.menuOpened = !this.menuOpened
    },
  },
}
</script>

