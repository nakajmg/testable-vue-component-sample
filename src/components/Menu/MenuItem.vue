<template>
  <div class="MenuItem">
    <span class="MenuItem_Label"
      @click="clickMenuItem"
    >
      {{label}}
    </span>
  </div>
</template>

<script>
export default {
  name: "MenuItem",
  props: {
    label: String,
    name: String,
  },
  methods: {
    clickMenuItem() {
      this.$emit("clickMenuItem", { name: this.name })
    },
  },
}
</script>

<style lang="scss">
.MenuItem {
  &_Label {
    cursor: pointer;
    &:hover {
      text-decoration: underline;
    }
  }
}
</style>
