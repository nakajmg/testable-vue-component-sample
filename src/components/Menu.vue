<template>
  <div class="Menu">
    <MenuItem class="Menu_Item"
      v-for="item in items"
      :key="item.label"
      v-bind="item"
      @clickMenuItem="onClickMenuItem"
    />
  </div>
</template>

<script>
import MenuItem from "../basics/MenuItem.vue"
export default {
  name: "Menu",
  components: {
    MenuItem,
  },
  props: {
    items: Array,
  },
  methods: {
    onClickMenuItem({ name }) {
      this.$emit("clickMenuItem", { name })
    },
  },
}
</script>

<style lang="scss">
.Menu {
  display: flex;
  align-items: center;
  &_Item + &_Item {
    margin-left: 1em;
  }
}
</style>
