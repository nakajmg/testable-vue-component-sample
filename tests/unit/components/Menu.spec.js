import { mount } from "@vue/test-utils"
import Menu from "@/components/Menu.vue"
import MenuItem from "@/basics/MenuItem.vue"
import menuItems from "../../_mockData/menuItems.json"
describe("Menu.vue", () => {
  const propsData = {
    items: menuItems,
  }
  it("props", () => {
    const wrapper = mount(Menu, { propsData })
    expect(wrapper.vm.$props).toEqual(propsData)
  })
  describe("methods", () => {
    it("clickMenuItem", () => {
      const wrapper = mount(Menu, { propsData })
      wrapper.vm.onClickMenuItem(menuItems[0])
      expect(wrapper.emitted("clickMenuItem")).toBeTruthy()
      expect(wrapper.emitted("clickMenuItem")[0][0]).toEqual({
        name: menuItems[0].name,
      })
    })
  })
  it("@clickMenuItem=onClickMenuItem", () => {
    const mock = jest.fn()
    const wrapper = mount(Menu, { propsData })
    wrapper.setMethods({
      onClickMenuItem: mock,
    })
    const menuItem = wrapper.find(MenuItem)
    menuItem.find(".MenuItem_Label").trigger("click")
    expect(mock).toHaveBeenCalledWith({
      name: menuItems[0].name,
    })
  })
  describe("template", () => {
    it("snapshot", () => {
      const wrapper = mount(Menu, { propsData })
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
  })
})
