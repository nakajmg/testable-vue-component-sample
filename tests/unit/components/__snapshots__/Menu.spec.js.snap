// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu.vue template snapshot 1`] = `
<div
  class="Menu"
>
  <div
    class="MenuItem Menu_Item"
  >
    <span
      class="MenuItem_Label"
    >
      
    About
  
    </span>
  </div>
  <div
    class="MenuItem Menu_Item"
  >
    <span
      class="MenuItem_Label"
    >
      
    Todos
  
    </span>
  </div>
</div>
`;
